File: lua/terra_sdk.lua
-- Terra SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Typed-model annotations (LuaLS ---@class); empty at runtime.
require("terra_types")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local TerraSDK = {}
TerraSDK.__index = TerraSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

TerraSDK._make_feature = _make_feature


function TerraSDK.new(options)
  local self = setmetatable({}, TerraSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

    -- feature: test


  return self
end


function TerraSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function TerraSDK:get_utility()
  return Utility.copy(self._utility)
end


function TerraSDK:get_root_ctx()
  return self._rootctx
end


function TerraSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function TerraSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Activity():list() / client:Activity():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Activity(data)
  local EntityMod = require("entity.activity_entity")
  if data == nil then
    if self._activity == nil then
      self._activity = EntityMod.new(self, nil)
    end
    return self._activity
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Athlete():list() / client:Athlete():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Athlete(data)
  local EntityMod = require("entity.athlete_entity")
  if data == nil then
    if self._athlete == nil then
      self._athlete = EntityMod.new(self, nil)
    end
    return self._athlete
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Authentication():list() / client:Authentication():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Authentication(data)
  local EntityMod = require("entity.authentication_entity")
  if data == nil then
    if self._authentication == nil then
      self._authentication = EntityMod.new(self, nil)
    end
    return self._authentication
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Body():list() / client:Body():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Body(data)
  local EntityMod = require("entity.body_entity")
  if data == nil then
    if self._body == nil then
      self._body = EntityMod.new(self, nil)
    end
    return self._body
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:BulkUserInfo():list() / client:BulkUserInfo():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:BulkUserInfo(data)
  local EntityMod = require("entity.bulk_user_info_entity")
  if data == nil then
    if self._bulk_user_info == nil then
      self._bulk_user_info = EntityMod.new(self, nil)
    end
    return self._bulk_user_info
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Daily():list() / client:Daily():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Daily(data)
  local EntityMod = require("entity.daily_entity")
  if data == nil then
    if self._daily == nil then
      self._daily = EntityMod.new(self, nil)
    end
    return self._daily
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Integration():list() / client:Integration():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Integration(data)
  local EntityMod = require("entity.integration_entity")
  if data == nil then
    if self._integration == nil then
      self._integration = EntityMod.new(self, nil)
    end
    return self._integration
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:LabReport():list() / client:LabReport():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:LabReport(data)
  local EntityMod = require("entity.lab_report_entity")
  if data == nil then
    if self._lab_report == nil then
      self._lab_report = EntityMod.new(self, nil)
    end
    return self._lab_report
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:LabReportDelivery():list() / client:LabReportDelivery():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:LabReportDelivery(data)
  local EntityMod = require("entity.lab_report_delivery_entity")
  if data == nil then
    if self._lab_report_delivery == nil then
      self._lab_report_delivery = EntityMod.new(self, nil)
    end
    return self._lab_report_delivery
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:LabReportFile():list() / client:LabReportFile():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:LabReportFile(data)
  local EntityMod = require("entity.lab_report_file_entity")
  if data == nil then
    if self._lab_report_file == nil then
      self._lab_report_file = EntityMod.new(self, nil)
    end
    return self._lab_report_file
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Menstruation():list() / client:Menstruation():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Menstruation(data)
  local EntityMod = require("entity.menstruation_entity")
  if data == nil then
    if self._menstruation == nil then
      self._menstruation = EntityMod.new(self, nil)
    end
    return self._menstruation
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Nutrition():list() / client:Nutrition():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Nutrition(data)
  local EntityMod = require("entity.nutrition_entity")
  if data == nil then
    if self._nutrition == nil then
      self._nutrition = EntityMod.new(self, nil)
    end
    return self._nutrition
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PlannedWorkout():list() / client:PlannedWorkout():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:PlannedWorkout(data)
  local EntityMod = require("entity.planned_workout_entity")
  if data == nil then
    if self._planned_workout == nil then
      self._planned_workout = EntityMod.new(self, nil)
    end
    return self._planned_workout
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Sleep():list() / client:Sleep():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Sleep(data)
  local EntityMod = require("entity.sleep_entity")
  if data == nil then
    if self._sleep == nil then
      self._sleep = EntityMod.new(self, nil)
    end
    return self._sleep
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:User():list() / client:User():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:User(data)
  local EntityMod = require("entity.user_entity")
  if data == nil then
    if self._user == nil then
      self._user = EntityMod.new(self, nil)
    end
    return self._user
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Workout():list() / client:Workout():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TerraSDK:Workout(data)
  local EntityMod = require("entity.workout_entity")
  if data == nil then
    if self._workout == nil then
      self._workout = EntityMod.new(self, nil)
    end
    return self._workout
  end
  return EntityMod.new(self, data)
end




function TerraSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = TerraSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return TerraSDK

File: php/terra_sdk.php
<?php
declare(strict_types=1);

// Terra SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

// Features record diagnostic state on the client as dynamic properties
// (_retry, _cache, _metrics, ...); allow them explicitly (PHP 8.2+
// deprecates implicit dynamic properties).
#[\AllowDynamicProperties]
class TerraSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new TerraUtility();
        $this->_utility = $utility;

        $config = TerraConfig::shared_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features in the resolved order (make_options puts an explicit
        // list order first, else defaults to test-first). Ordering matters: the
        // `test` feature installs the base mock transport and the transport
        // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        // current, so `test` must be added before them to sit at the base.
        $feature_opts = TerraHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $featureorder = Struct::getpath($this->options, "__derived__.featureorder");
            if (is_array($featureorder)) {
                foreach ($featureorder as $fname) {
                    $fopts = TerraHelpers::to_map($feature_opts[$fname] ?? null);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, TerraFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return TerraUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = TerraHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = TerraHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = TerraHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new TerraSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    // Raw endpoint access is operator-controllable, like every entity op.
    // Blocking it means denying BOTH the 'direct' and 'graphql' tokens,
    // since either one reaches the same endpoint.
    public function direct(array $fetchargs = []): mixed
    {
        if (!$this->op_allowed("direct")) {
            return $this->op_denied("direct");
        }

        return $this->raw_request($fetchargs);
    }

    // Is this raw-access op permitted by the SDK's allow.op option?
    private function op_allowed(string $op): bool
    {
        $allow_op = Struct::getpath($this->options, "allow.op");
        return is_string($allow_op) && str_contains($allow_op, $op);
    }

    private function op_denied(string $op): array
    {
        $allow_op = Struct::getpath($this->options, "allow.op");
        return [
            "ok" => false,
            "err" => new TerraError($op . "_allow",
                "TerraSDK: " . $op . ": operation not allowed by" .
                " SDK option allow.op value: \"" . (string)$allow_op . "\""),
        ];
    }

    // Ungated request path shared by direct and graphql, each of which
    // checks its own allow.op token first. Private, rather than a flag on
    // fetchargs: a caller-supplied marker would let anyone opt straight back
    // out of the gate by passing it.
    private function raw_request(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = TerraHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = TerraHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }

    // Raw GraphQL access: the pressure valve that makes the generated
    // surface's deliberate omissions (per-call selection sets, typed filter
    // builders, batching, subscriptions) livable — the whole schema stays
    // reachable.
    //
    // Thin wrapper over the same prepare/fetch path direct uses, with the
    // one thing raw direct cannot do for GraphQL: a GraphQL failure rides
    // HTTP 200 as a top-level `errors` array, so status alone would report
    // a failed query as ok.
    //
    // NOTE: like direct, this bypasses the feature pipeline — no retry,
    // ratelimit or paging features apply.
    public function graphql(string $query, ?array $variables = null, ?array $ctrl = null): mixed
    {
        if (!$this->op_allowed("graphql")) {
            return $this->op_denied("graphql");
        }

        $res = $this->raw_request([
            "method" => "POST",
            "headers" => ["content-type" => "application/json"],
            "body" => ["query" => $query, "variables" => $variables ?? []],
            "ctrl" => $ctrl ?? [],
        ]);

        if (!is_array($res)) {
            return $res;
        }

        // Errors are read BEFORE any status check: a GraphQL parse or
        // validation failure comes back as HTTP 400 carrying the standard
        // { errors: [...] } body, and the raw path represents a non-2xx as
        // ok:false with no err — so returning early on status would discard
        // the server's own diagnostics, which are the only useful part of
        // that response.
        $errors = Struct::getpath($res, "data.errors");

        if (is_array($errors) && 0 < count($errors)) {
            $first = is_array($errors[0]) ? $errors[0] : [];
            $msg = $first["message"] ?? "";
            if (!is_string($msg) || "" === $msg) {
                $msg = "graphql error";
            }
            $res["ok"] = false;
            $res["err"] = new TerraError("graphql_error",
                "TerraSDK: graphql: " . $msg);
            $res["graphql"] = $errors;
        }

        return $res;
    }


    private $_activity = null;

    // Canonical facade: $client->Activity()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->activity()
    // resolves here too.
    public function Activity($data = null)
    {
        require_once __DIR__ . '/entity/activity_entity.php';
        if ($data === null) {
            if ($this->_activity === null) {
                $this->_activity = new ActivityEntity($this, null);
            }
            return $this->_activity;
        }
        return new ActivityEntity($this, $data);
    }


    private $_athlete = null;

    // Canonical facade: $client->Athlete()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->athlete()
    // resolves here too.
    public function Athlete($data = null)
    {
        require_once __DIR__ . '/entity/athlete_entity.php';
        if ($data === null) {
            if ($this->_athlete === null) {
                $this->_athlete = new AthleteEntity($this, null);
            }
            return $this->_athlete;
        }
        return new AthleteEntity($this, $data);
    }


    private $_authentication = null;

    // Canonical facade: $client->Authentication()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->authentication()
    // resolves here too.
    public function Authentication($data = null)
    {
        require_once __DIR__ . '/entity/authentication_entity.php';
        if ($data === null) {
            if ($this->_authentication === null) {
                $this->_authentication = new AuthenticationEntity($this, null);
            }
            return $this->_authentication;
        }
        return new AuthenticationEntity($this, $data);
    }


    private $_body = null;

    // Canonical facade: $client->Body()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->body()
    // resolves here too.
    public function Body($data = null)
    {
        require_once __DIR__ . '/entity/body_entity.php';
        if ($data === null) {
            if ($this->_body === null) {
                $this->_body = new BodyEntity($this, null);
            }
            return $this->_body;
        }
        return new BodyEntity($this, $data);
    }


    private $_bulk_user_info = null;

    // Canonical facade: $client->BulkUserInfo()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->bulk_user_info()
    // resolves here too.
    public function BulkUserInfo($data = null)
    {
        require_once __DIR__ . '/entity/bulk_user_info_entity.php';
        if ($data === null) {
            if ($this->_bulk_user_info === null) {
                $this->_bulk_user_info = new BulkUserInfoEntity($this, null);
            }
            return $this->_bulk_user_info;
        }
        return new BulkUserInfoEntity($this, $data);
    }


    private $_daily = null;

    // Canonical facade: $client->Daily()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->daily()
    // resolves here too.
    public function Daily($data = null)
    {
        require_once __DIR__ . '/entity/daily_entity.php';
        if ($data === null) {
            if ($this->_daily === null) {
                $this->_daily = new DailyEntity($this, null);
            }
            return $this->_daily;
        }
        return new DailyEntity($this, $data);
    }


    private $_integration = null;

    // Canonical facade: $client->Integration()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->integration()
    // resolves here too.
    public function Integration($data = null)
    {
        require_once __DIR__ . '/entity/integration_entity.php';
        if ($data === null) {
            if ($this->_integration === null) {
                $this->_integration = new IntegrationEntity($this, null);
            }
            return $this->_integration;
        }
        return new IntegrationEntity($this, $data);
    }


    private $_lab_report = null;

    // Canonical facade: $client->LabReport()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->lab_report()
    // resolves here too.
    public function LabReport($data = null)
    {
        require_once __DIR__ . '/entity/lab_report_entity.php';
        if ($data === null) {
            if ($this->_lab_report === null) {
                $this->_lab_report = new LabReportEntity($this, null);
            }
            return $this->_lab_report;
        }
        return new LabReportEntity($this, $data);
    }


    private $_lab_report_delivery = null;

    // Canonical facade: $client->LabReportDelivery()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->lab_report_delivery()
    // resolves here too.
    public function LabReportDelivery($data = null)
    {
        require_once __DIR__ . '/entity/lab_report_delivery_entity.php';
        if ($data === null) {
            if ($this->_lab_report_delivery === null) {
                $this->_lab_report_delivery = new LabReportDeliveryEntity($this, null);
            }
            return $this->_lab_report_delivery;
        }
        return new LabReportDeliveryEntity($this, $data);
    }


    private $_lab_report_file = null;

    // Canonical facade: $client->LabReportFile()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->lab_report_file()
    // resolves here too.
    public function LabReportFile($data = null)
    {
        require_once __DIR__ . '/entity/lab_report_file_entity.php';
        if ($data === null) {
            if ($this->_lab_report_file === null) {
                $this->_lab_report_file = new LabReportFileEntity($this, null);
            }
            return $this->_lab_report_file;
        }
        return new LabReportFileEntity($this, $data);
    }


    private $_menstruation = null;

    // Canonical facade: $client->Menstruation()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->menstruation()
    // resolves here too.
    public function Menstruation($data = null)
    {
        require_once __DIR__ . '/entity/menstruation_entity.php';
        if ($data === null) {
            if ($this->_menstruation === null) {
                $this->_menstruation = new MenstruationEntity($this, null);
            }
            return $this->_menstruation;
        }
        return new MenstruationEntity($this, $data);
    }


    private $_nutrition = null;

    // Canonical facade: $client->Nutrition()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->nutrition()
    // resolves here too.
    public function Nutrition($data = null)
    {
        require_once __DIR__ . '/entity/nutrition_entity.php';
        if ($data === null) {
            if ($this->_nutrition === null) {
                $this->_nutrition = new NutritionEntity($this, null);
            }
            return $this->_nutrition;
        }
        return new NutritionEntity($this, $data);
    }


    private $_planned_workout = null;

    // Canonical facade: $client->PlannedWorkout()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->planned_workout()
    // resolves here too.
    public function PlannedWorkout($data = null)
    {
        require_once __DIR__ . '/entity/planned_workout_entity.php';
        if ($data === null) {
            if ($this->_planned_workout === null) {
                $this->_planned_workout = new PlannedWorkoutEntity($this, null);
            }
            return $this->_planned_workout;
        }
        return new PlannedWorkoutEntity($this, $data);
    }


    private $_sleep = null;

    // Canonical facade: $client->Sleep()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->sleep()
    // resolves here too.
    public function Sleep($data = null)
    {
        require_once __DIR__ . '/entity/sleep_entity.php';
        if ($data === null) {
            if ($this->_sleep === null) {
                $this->_sleep = new SleepEntity($this, null);
            }
            return $this->_sleep;
        }
        return new SleepEntity($this, $data);
    }


    private $_user = null;

    // Canonical facade: $client->User()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->user()
    // resolves here too.
    public function User($data = null)
    {
        require_once __DIR__ . '/entity/user_entity.php';
        if ($data === null) {
            if ($this->_user === null) {
                $this->_user = new UserEntity($this, null);
            }
            return $this->_user;
        }
        return new UserEntity($this, $data);
    }


    private $_workout = null;

    // Canonical facade: $client->Workout()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->workout()
    // resolves here too.
    public function Workout($data = null)
    {
        require_once __DIR__ . '/entity/workout_entity.php';
        if ($data === null) {
            if ($this->_workout === null) {
                $this->_workout = new WorkoutEntity($this, null);
            }
            return $this->_workout;
        }
        return new WorkoutEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new TerraSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}
